[Bug] Demo Flink SQL can not be run on yarn-per-job mode with flink 1.16 and flink 1.17

### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-streampark/issues?q=is%3Aissue+label%3A%22bug%22) and found no similar issues.


### Java Version

1.8

### Scala Version

2.12.x

### StreamPark Version

dev

### Flink Version

1.16
1.17

### deploy mode

yarn-perjob

### What happened
I have tested that shim's packaged `streampark-flinkjob_{applicationName}.jar` running properly locally using `flink run -t yarn-per-job --detached`. But it can not be run in streampark.

```
java.util.concurrent.CompletionException: java.lang.reflect.InvocationTargetException
	at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:273)
	at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:280)
	at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1606)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:750)
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.streampark.flink.client.FlinkClient$.$anonfun$proxy$1(FlinkClient.scala:88)
	at org.apache.streampark.flink.proxy.FlinkShimsProxy$.$anonfun$proxy$1(FlinkShimsProxy.scala:62)
	at org.apache.streampark.common.util.ClassLoaderUtils$.runAsClassLoader(ClassLoaderUtils.scala:42)
	at org.apache.streampark.flink.proxy.FlinkShimsProxy$.proxy(FlinkShimsProxy.scala:62)
	at org.apache.streampark.flink.client.FlinkClient$.proxy(FlinkClient.scala:83)
	at org.apache.streampark.flink.client.FlinkClient$.submit(FlinkClient.scala:54)
	at org.apache.streampark.flink.client.FlinkClient.submit(FlinkClient.scala)
	at org.apache.streampark.console.core.service.application.impl.ApplicationActionServiceImpl.lambda$start$2(ApplicationActionServiceImpl.java:464)
	at java.util.concurrent.CompletableFuture$AsyncSupply.run(CompletableFuture.java:1604)
	... 3 more
Caused by: org.apache.flink.client.program.ProgramInvocationException: The main method caused an error: Unable to instantiate java compiler
	at org.apache.flink.client.program.PackagedProgram.callMainMethod(PackagedProgram.java:372)
	at org.apache.flink.client.program.PackagedProgram.invokeInteractiveModeForExecution(PackagedProgram.java:222)
	at org.apache.flink.client.program.PackagedProgramUtils.getPipelineFromProgram(PackagedProgramUtils.java:158)
	at org.apache.flink.client.program.PackagedProgramUtils.createJobGraph(PackagedProgramUtils.java:82)
	at org.apache.streampark.flink.client.trait.FlinkClientTrait.getJobGraph(FlinkClientTrait.scala:293)
	at org.apache.streampark.flink.client.trait.FlinkClientTrait.getJobGraph$(FlinkClientTrait.scala:251)
	at org.apache.streampark.flink.client.impl.YarnPerJobClient$.doSubmit(YarnPerJobClient.scala:86)
	at org.apache.streampark.flink.client.trait.FlinkClientTrait.submit(FlinkClientTrait.scala:158)
	at org.apache.streampark.flink.client.trait.FlinkClientTrait.submit$(FlinkClientTrait.scala:70)
	at org.apache.streampark.flink.client.impl.YarnPerJobClient$.submit(YarnPerJobClient.scala:40)
	at org.apache.streampark.flink.client.FlinkClientEntrypoint$.submit(FlinkClientEntrypoint.scala:47)
	at org.apache.streampark.flink.client.FlinkClientEntrypoint.submit(FlinkClientEntrypoint.scala)
	... 16 more
Caused by: java.lang.IllegalStateException: Unable to instantiate java compiler
	at org.apache.calcite.rel.metadata.JaninoRelMetadataProvider.compile(JaninoRelMetadataProvider.java:163)
	at org.apache.calcite.rel.metadata.JaninoRelMetadataProvider.generateCompileAndInstantiate(JaninoRelMetadataProvider.java:141)
	at org.apache.calcite.rel.metadata.JaninoRelMetadataProvider.lambda$static$0(JaninoRelMetadataProvider.java:73)
	at org.apache.flink.calcite.shaded.com.google.common.cache.CacheLoader$FunctionToCacheLoader.load(CacheLoader.java:165)
	at org.apache.flink.calcite.shaded.com.google.common.cache.LocalCache$LoadingValueReference.loadFuture(LocalCache.java:3529)
	at org.apache.flink.calcite.shaded.com.google.common.cache.LocalCache$Segment.loadSync(LocalCache.java:2278)
	at org.apache.flink.calcite.shaded.com.google.common.cache.LocalCache$Segment.lockedGetOrLoad(LocalCache.java:2155)
	at org.apache.flink.calcite.shaded.com.google.common.cache.LocalCache$Segment.get(LocalCache.java:2045)
	at org.apache.flink.calcite.shaded.com.google.common.cache.LocalCache.get(LocalCache.java:3951)
	at org.apache.flink.calcite.shaded.com.google.common.cache.LocalCache.getOrLoad(LocalCache.java:3974)
	at org.apache.flink.calcite.shaded.com.google.common.cache.LocalCache$LocalLoadingCache.get(LocalCache.java:4958)
	at org.apache.calcite.rel.metadata.JaninoRelMetadataProvider.revise(JaninoRelMetadataProvider.java:197)
	at org.apache.calcite.rel.metadata.RelMetadataQueryBase.revise(RelMetadataQueryBase.java:118)
	at org.apache.calcite.rel.metadata.RelMetadataQuery.getPulledUpPredicates(RelMetadataQuery.java:844)
	at org.apache.calcite.rel.rules.ReduceExpressionsRule$ProjectReduceExpressionsRule.onMatch(ReduceExpressionsRule.java:307)
	at org.apache.calcite.plan.AbstractRelOptPlanner.fireRule(AbstractRelOptPlanner.java:337)
	at org.apache.calcite.plan.hep.HepPlanner.applyRule(HepPlanner.java:565)
	at org.apache.calcite.plan.hep.HepPlanner.applyRules(HepPlanner.java:428)
	at org.apache.calcite.plan.hep.HepPlanner.executeInstruction(HepPlanner.java:251)
	at org.apache.calcite.plan.hep.HepInstruction$RuleInstance.execute(HepInstruction.java:130)
	at org.apache.calcite.plan.hep.HepPlanner.executeProgram(HepPlanner.java:208)
	at org.apache.calcite.plan.hep.HepPlanner.findBestExp(HepPlanner.java:195)
	at org.apache.flink.table.planner.plan.optimize.program.FlinkHepProgram.optimize(FlinkHepProgram.scala:64)
	at org.apache.flink.table.planner.plan.optimize.program.FlinkHepRuleSetProgram.optimize(FlinkHepRuleSetProgram.scala:78)
	at org.apache.flink.table.planner.plan.optimize.program.FlinkChainedProgram.$anonfun$optimize$1(FlinkChainedProgram.scala:59)
	at scala.collection.TraversableOnce.$anonfun$foldLeft$1(TraversableOnce.scala:156)
	at scala.collection.TraversableOnce.$anonfun$foldLeft$1$adapted(TraversableOnce.scala:156)
	at scala.collection.Iterator.foreach(Iterator.scala:937)
	at scala.collection.Iterator.foreach$(Iterator.scala:937)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1425)
	at scala.collection.IterableLike.foreach(IterableLike.scala:70)
	at scala.collection.IterableLike.foreach$(IterableLike.scala:69)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:54)
	at scala.collection.TraversableOnce.foldLeft(TraversableOnce.scala:156)
	at scala.collection.TraversableOnce.foldLeft$(TraversableOnce.scala:154)
	at scala.collection.AbstractTraversable.foldLeft(Traversable.scala:104)
	at org.apache.flink.table.planner.plan.optimize.program.FlinkChainedProgram.optimize(FlinkChainedProgram.scala:55)
	at org.apache.flink.table.planner.plan.optimize.StreamCommonSubGraphBasedOptimizer.optimizeTree(StreamCommonSubGraphBasedOptimizer.scala:176)
	at org.apache.flink.table.planner.plan.optimize.StreamCommonSubGraphBasedOptimizer.doOptimize(StreamCommonSubGraphBasedOptimizer.scala:83)
	at org.apache.flink.table.planner.plan.optimize.CommonSubGraphBasedOptimizer.optimize(CommonSubGraphBasedOptimizer.scala:87)
	at org.apache.flink.table.planner.delegation.PlannerBase.optimize(PlannerBase.scala:329)
	at org.apache.flink.table.planner.delegation.PlannerBase.translate(PlannerBase.scala:195)
	at org.apache.flink.table.api.internal.TableEnvironmentImpl.translate(TableEnvironmentImpl.java:1805)
	at org.apache.flink.table.api.internal.TableEnvironmentImpl.executeInternal(TableEnvironmentImpl.java:881)
	at org.apache.flink.table.api.internal.StatementSetImpl.execute(StatementSetImpl.java:109)
	at org.apache.streampark.flink.core.FlinkSqlExecutor$.executeSql(FlinkSqlExecutor.scala:148)
	at org.apache.streampark.flink.core.FlinkStreamTableTrait.sql(FlinkStreamTableTrait.scala:88)
	at org.apache.streampark.flink.cli.SqlClient$StreamSqlApp$.handle(SqlClient.scala:90)
	at org.apache.streampark.flink.core.scala.FlinkStreamTable.main(FlinkStreamTable.scala:48)
	at org.apache.streampark.flink.core.scala.FlinkStreamTable.main$(FlinkStreamTable.scala:45)
	at org.apache.streampark.flink.cli.SqlClient$StreamSqlApp$.main(SqlClient.scala:89)
	at org.apache.streampark.flink.cli.SqlClient$.delayedEndpoint$org$apache$streampark$flink$cli$SqlClient$1(SqlClient.scala:78)
	at org.apache.streampark.flink.cli.SqlClient$delayedInit$body.apply(SqlClient.scala:34)
	at scala.Function0.apply$mcV$sp(Function0.scala:34)
	at scala.Function0.apply$mcV$sp$(Function0.scala:34)
	at scala.runtime.AbstractFunction0.apply$mcV$sp(AbstractFunction0.scala:12)
	at scala.App.$anonfun$main$1$adapted(App.scala:76)
	at scala.collection.immutable.List.foreach(List.scala:388)
	at scala.App.main(App.scala:76)
	at scala.App.main$(App.scala:74)
	at org.apache.streampark.flink.cli.SqlClient$.main(SqlClient.scala:34)
	at org.apache.streampark.flink.cli.SqlClient.main(SqlClient.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.apache.flink.client.program.PackagedProgram.callMainMethod(PackagedProgram.java:355)
	... 27 more
Caused by: java.lang.ClassCastException: org.codehaus.janino.CompilerFactory cannot be cast to org.codehaus.commons.compiler.ICompilerFactory
	at org.codehaus.commons.compiler.CompilerFactoryFactory.getCompilerFactory(CompilerFactoryFactory.java:129)
	at org.codehaus.commons.compiler.CompilerFactoryFactory.getDefaultCompilerFactory(CompilerFactoryFactory.java:79)
	at org.apache.calcite.rel.metadata.JaninoRelMetadataProvider.compile(JaninoRelMetadataProvider.java:161)
	... 93 more
```

### Error Exception

_No response_

### Screenshots

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!(您是否要贡献这个PR?)

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

--- a/streampark-e2e/streampark-e2e-case/src/test/java/org/apache/streampark/e2e/cases/ApplicationsFlink116OnYarnWithFlinkSQLTest.java
+++ b/streampark-e2e/streampark-e2e-case/src/test/java/org/apache/streampark/e2e/cases/ApplicationsFlink116OnYarnWithFlinkSQLTest.java
@@ -213,9 +213,8 @@ public class ApplicationsFlink116OnYarnWithFlinkSQLTest {
                                 .anyMatch(it -> it.contains("SUCCESS")));
     }
 
-    // This test cannot be executed due to a bug, and will be put online after issue #3761 fixed
-    // @Test
-    // @Order(70)
+    @Test
+    @Order(70)
     void testStartFlinkApplicationOnYarnPerJobMode() {
         final ApplicationsPage applicationsPage = new ApplicationsPage(browser);
 

--- a/streampark-e2e/streampark-e2e-case/src/test/java/org/apache/streampark/e2e/cases/ApplicationsFlink117OnYarnWithFlinkSQLTest.java
+++ b/streampark-e2e/streampark-e2e-case/src/test/java/org/apache/streampark/e2e/cases/ApplicationsFlink117OnYarnWithFlinkSQLTest.java
@@ -214,9 +214,8 @@ public class ApplicationsFlink117OnYarnWithFlinkSQLTest {
                                 .anyMatch(it -> it.contains("SUCCESS")));
     }
 
-    // This test cannot be executed due to a bug, and will be put online after issue #3761 fixed
-    // @Test
-    // @Order(70)
+    @Test
+    @Order(70)
     void testStartFlinkApplicationOnYarnPerJobMode() {
         final ApplicationsPage applicationsPage = new ApplicationsPage(browser);
 

--- a/streampark-flink/streampark-flink-client/streampark-flink-client-api/src/main/scala/org/apache/streampark/flink/client/bean/SubmitRequest.scala
+++ b/streampark-flink/streampark-flink-client/streampark-flink-client-api/src/main/scala/org/apache/streampark/flink/client/bean/SubmitRequest.scala
@@ -21,7 +21,7 @@ import org.apache.streampark.common.Constant
 import org.apache.streampark.common.conf.{FlinkVersion, Workspace}
 import org.apache.streampark.common.conf.ConfigKeys._
 import org.apache.streampark.common.enums._
-import org.apache.streampark.common.util.{AssertUtils, DeflaterUtils, HdfsUtils, PropertiesUtils}
+import org.apache.streampark.common.util._
 import org.apache.streampark.flink.packer.pipeline.{BuildResult, ShadedBuildResponse}
 import org.apache.streampark.flink.util.FlinkUtils
 import org.apache.streampark.shaded.com.fasterxml.jackson.databind.ObjectMapper
@@ -110,6 +110,8 @@ case class SubmitRequest(
       case _ => false
     }
   }
+
+  def hasProp(key: String): Boolean = properties.containsKey(key)
 
   private[this] def getParameterMap(prefix: String = ""): Map[String, String] = {
     if (this.appConf == null) {

--- a/streampark-flink/streampark-flink-client/streampark-flink-client-core/src/main/scala/org/apache/streampark/flink/client/trait/FlinkClientTrait.scala
+++ b/streampark-flink/streampark-flink-client/streampark-flink-client-core/src/main/scala/org/apache/streampark/flink/client/trait/FlinkClientTrait.scala
@@ -20,10 +20,10 @@ package org.apache.streampark.flink.client.`trait`
 import org.apache.streampark.common.Constant
 import org.apache.streampark.common.conf.ConfigKeys._
 import org.apache.streampark.common.conf.Workspace
-import org.apache.streampark.common.enums._
+import org.apache.streampark.common.enums.{ApplicationType, FlinkDevelopmentMode, FlinkExecutionMode, FlinkRestoreMode}
 import org.apache.streampark.common.fs.FsOperator
-import org.apache.streampark.common.util._
-import org.apache.streampark.flink.client.bean.{SubmitResponse, _}
+import org.apache.streampark.common.util.{AssertUtils, DeflaterUtils, ExceptionUtils, Logger, PropertiesUtils, SystemPropertyUtils, Utils}
+import org.apache.streampark.flink.client.bean._
 import org.apache.streampark.flink.core.FlinkClusterClient
 import org.apache.streampark.flink.core.conf.FlinkRunOption
 
@@ -39,7 +39,6 @@ import org.apache.flink.client.program.{ClusterClient, PackagedProgram, Packaged
 import org.apache.flink.configuration._
 import org.apache.flink.python.PythonOptions
 import org.apache.flink.runtime.jobgraph.{JobGraph, SavepointConfigOptions}
-import org.apache.flink.util.FlinkException
 import org.apache.flink.util.Preconditions.checkNotNull
 
 import java.io.File
@@ -145,8 +144,9 @@ trait FlinkClientTrait extends Logger {
       case Success(resp) => resp
       case Failure(e) =>
         logError(
-          s"flink job ${submitRequest.appName} start failed, executionMode: ${submitRequest.executionMode.getName}, detail: ${ExceptionUtils
-              .stringifyException(e)}")
+          s"flink job ${submitRequest.appName} start failed, " +
+            s"executionMode: ${submitRequest.executionMode.getName}, " +
+            s"detail: ${ExceptionUtils.stringifyException(e)}")
         throw e
     }
   }
@@ -262,39 +262,46 @@ trait FlinkClientTrait extends Logger {
       submitRequest: SubmitRequest,
       jarFile: File): (PackagedProgram, JobGraph) = {
 
-    val pkgBuilder = PackagedProgram.newBuilder
-      .setEntryPointClassName(
-        flinkConfig.getOptional(ApplicationConfiguration.APPLICATION_MAIN_CLASS).get()
-      )
-      .setArguments(
-        flinkConfig
-          .getOptional(ApplicationConfiguration.APPLICATION_ARGS)
-          .orElse(Lists.newArrayList()): _*
-      )
-      .setSavepointRestoreSettings(submitRequest.savepointRestoreSettings)
+    val packagedProgramBuilder = {
+      val builder = PackagedProgram.newBuilder
+        .setSavepointRestoreSettings(submitRequest.savepointRestoreSettings)
+        .setEntryPointClassName(
+          flinkConfig.getOptional(ApplicationConfiguration.APPLICATION_MAIN_CLASS).get()
+        )
+        .setArguments(
+          flinkConfig
+            .getOptional(ApplicationConfiguration.APPLICATION_ARGS)
+            .orElse(Lists.newArrayList()): _*
+        )
 
-    submitRequest.developmentMode match {
-      case FlinkDevelopmentMode.PYFLINK =>
-        val pythonVenv: String = Workspace.local.APP_PYTHON_VENV
-        AssertUtils.required(FsOperator.lfs.exists(pythonVenv), s"$pythonVenv File does not exist")
+      submitRequest.developmentMode match {
+        case FlinkDevelopmentMode.PYFLINK =>
+          val pythonVenv: String = Workspace.local.APP_PYTHON_VENV
 
-        flinkConfig
-          // python.archives
-          .safeSet(PythonOptions.PYTHON_ARCHIVES, pythonVenv)
-          // python.client.executable
-          .safeSet(PythonOptions.PYTHON_CLIENT_EXECUTABLE, Constant.PYTHON_EXECUTABLE)
-          // python.executable
-          .safeSet(PythonOptions.PYTHON_EXECUTABLE, Constant.PYTHON_EXECUTABLE)
-        if (submitRequest.libs.nonEmpty) {
-          pkgBuilder.setUserClassPaths(submitRequest.libs)
-        }
-      case _ =>
-        pkgBuilder
-          .setUserClassPaths(submitRequest.classPaths)
-          .setJarFile(jarFile)
+          AssertUtils.required(
+            FsOperator.lfs.exists(pythonVenv),
+            s"$pythonVenv File does not exist")
+          flinkConfig
+            // python.archives
+            .safeSet(PythonOptions.PYTHON_ARCHIVES, pythonVenv)
+            // python.client.executable
+            .safeSet(PythonOptions.PYTHON_CLIENT_EXECUTABLE, Constant.PYTHON_EXECUTABLE)
+            // python.executable
+            .safeSet(PythonOptions.PYTHON_EXECUTABLE, Constant.PYTHON_EXECUTABLE)
+          if (submitRequest.libs.nonEmpty) {
+            // BUG: https://github.com/apache/incubator-streampark/issues/3761
+            // builder.setUserClassPaths(Lists.newArrayList(submitRequest.libs: _*))
+          }
+        case _ =>
+          builder
+            // BUG: https://github.com/apache/incubator-streampark/issues/3761
+            // .setUserClassPaths(Lists.newArrayList(submitRequest.classPaths: _*))
+            .setJarFile(jarFile)
+      }
+      builder
     }
 
-    val packageProgram = pkgBuilder.build()
+    val packageProgram = packagedProgramBuilder.build()
     val jobGraph = PackagedProgramUtils.createJobGraph(
       packageProgram,
       flinkConfig,
@@ -339,10 +346,7 @@ trait FlinkClientTrait extends Logger {
     // 1. find the configuration directory
     val configurationDirectory = s"$flinkHome/conf"
     // 2. load the custom command lines
-    val customCommandLines =
-      loadCustomCommandLines(flinkDefaultConfiguration, configurationDirectory)
-    new CliFrontend(flinkDefaultConfiguration, customCommandLines)
-    customCommandLines
+    loadCustomCommandLines(flinkDefaultConfiguration, configurationDirectory)
   }
 
   private[client] def getParallelism(submitRequest: SubmitRequest): Integer = {
@@ -363,21 +367,25 @@ trait FlinkClientTrait extends Logger {
     val cliArgs = {
       val optionMap = new mutable.HashMap[String, Any]()
       submitRequest.appOption
-        .filter(
-          x => {
-            val verify = commandLineOptions.hasOption(x._1)
-            if (!verify) logWarn(s"param:${x._1} is error,skip it.")
-            verify
-          })
-        .foreach(
-          x => {
-            val opt = commandLineOptions.getOption(x._1.trim).getOpt
-            Try(x._2.toBoolean).getOrElse(x._2) match {
-              case b if b.isInstanceOf[Boolean] =>
-                if (b.asInstanceOf[Boolean]) optionMap += s"-$opt" -> true
-              case v => optionMap += s"-$opt" -> v
+        .filter {
+          opt =>
+            val verify = commandLineOptions.hasOption(opt._1)
+            if (!verify) {
+              logWarn(s"param:${opt._1} is error,skip it.")
             }
-          })
+            verify
+        }
+        .foreach {
+          opt =>
+            val option = commandLineOptions.getOption(opt._1.trim).getOpt
+            Try(opt._2.toBoolean).getOrElse(opt._2) match {
+              case b if b.isInstanceOf[Boolean] =>
+                if (b.asInstanceOf[Boolean]) {
+                  optionMap += s"-$option" -> true
+                }
+              case v => optionMap += s"-$option" -> v
+            }
+        }
 
       // fromSavePoint
       if (submitRequest.savePoint != null) {
@@ -391,9 +399,9 @@ trait FlinkClientTrait extends Logger {
 
       val array = new ArrayBuffer[String]()
       optionMap.foreach(
-        x => {
-          array += x._1
-          x._2 match {
+        opt => {
+          array += opt._1
+          opt._2 match {
             case v: String => array += v
             case _ =>
           }
@@ -401,12 +409,13 @@ trait FlinkClientTrait extends Logger {
 
       // app properties
       if (MapUtils.isNotEmpty(submitRequest.properties)) {
-        submitRequest.properties.foreach(
-          x => {
-            if (!x._1.startsWith(CoreOptions.FLINK_JVM_OPTIONS.key())) {
-              array += s"-D${x._1}=${x._2}"
+        submitRequest.properties.foreach {
+          key =>
+            if (!key._1.startsWith(CoreOptions.FLINK_JVM_OPTIONS.key())) {
+              logInfo(s"submit application dynamicProperties:  ${key._1} :${key._2}")
+              array += s"-D${key._1}=${key._2}"
             }
-          })
+        }
       }
       array.toArray
     }
@@ -480,21 +489,22 @@ trait FlinkClientTrait extends Logger {
     }
 
     // execution.runtime-mode
-    val addRuntimeModeState =
-      submitRequest.properties.nonEmpty && submitRequest.properties.containsKey(
-        ExecutionOptions.RUNTIME_MODE.key())
-    if (addRuntimeModeState) {
-      programArgs += s"--${ExecutionOptions.RUNTIME_MODE.key()}"
-      programArgs += submitRequest.properties.get(ExecutionOptions.RUNTIME_MODE.key()).toString
+    Try(submitRequest.properties(ExecutionOptions.RUNTIME_MODE.key()).toString) match {
+      case Success(runtimeMode) =>
+        programArgs += s"--${ExecutionOptions.RUNTIME_MODE.key()}"
+        programArgs += runtimeMode
+      case _ =>
     }
 
-    val addUserJarFileState =
-      submitRequest.developmentMode == FlinkDevelopmentMode.PYFLINK && submitRequest.executionMode != FlinkExecutionMode.YARN_APPLICATION
-    if (addUserJarFileState) {
-      // python file
-      programArgs.add("-py")
-      programArgs.add(submitRequest.userJarFile.getAbsolutePath)
+    if (submitRequest.developmentMode == FlinkDevelopmentMode.PYFLINK) {
+      // TODO why executionMode is not yarn-application ???
+      if (submitRequest.executionMode != FlinkExecutionMode.YARN_APPLICATION) {
+        // python file
+        programArgs.add("-py")
+        programArgs.add(submitRequest.userJarFile.getAbsolutePath)
+      }
     }
+
     Lists.newArrayList(programArgs: _*)
   }
 
@@ -507,9 +517,9 @@ trait FlinkClientTrait extends Logger {
     val configuration = new Configuration()
     val flinkDefaultConfiguration = getFlinkDefaultConfiguration(flinkHome)
     flinkDefaultConfiguration.keySet.foreach(
-      x => {
-        flinkDefaultConfiguration.getString(x, null) match {
-          case v if v != null => configuration.setString(x, v)
+      key => {
+        flinkDefaultConfiguration.getString(key, null) match {
+          case v if v != null => configuration.setString(key, v)
           case _ =>
         }
       })
@@ -537,9 +547,7 @@ trait FlinkClientTrait extends Logger {
     val withSavepoint = Try(cancelRequest.withSavepoint).getOrElse(false)
     val withDrain = Try(cancelRequest.withDrain).getOrElse(false)
 
-    (
-      Try(cancelRequest.withSavepoint).getOrElse(false),
-      Try(cancelRequest.withDrain).getOrElse(false)) match {
+    (withSavepoint, withDrain) match {
       case (false, false) =>
         client.cancel(jobID).get()
         null

--- a/streampark-flink/streampark-flink-client/streampark-flink-client-core/src/main/scala/org/apache/streampark/flink/client/trait/FlinkClientTrait.scala
+++ b/streampark-flink/streampark-flink-client/streampark-flink-client-core/src/main/scala/org/apache/streampark/flink/client/trait/FlinkClientTrait.scala
@@ -144,9 +144,7 @@ trait FlinkClientTrait extends Logger {
       .safeSet(ApplicationConfiguration.APPLICATION_ARGS, extractProgramArgs(submitRequest))
       .safeSet(PipelineOptionsInternal.PIPELINE_FIXED_JOB_ID, submitRequest.jobId)
 
-    if (
-      !submitRequest.properties.containsKey(CheckpointingOptions.MAX_RETAINED_CHECKPOINTS.key())
-    ) {
+    if (!submitRequest.hasProp(CheckpointingOptions.MAX_RETAINED_CHECKPOINTS.key())) {
       val flinkDefaultConfiguration = getFlinkDefaultConfiguration(
         submitRequest.flinkVersion.flinkHome)
       // state.checkpoints.num-retained
@@ -160,9 +158,10 @@ trait FlinkClientTrait extends Logger {
       flinkConfig.setBoolean(
         SavepointConfigOptions.SAVEPOINT_IGNORE_UNCLAIMED_STATE,
         submitRequest.allowNonRestoredState)
-      val enableRestoreModeState = submitRequest.flinkVersion.checkVersion(
-        FlinkRestoreMode.SINCE_FLINK_VERSION) && submitRequest.restoreMode != null
-      if (enableRestoreModeState) {
+      val enableRestoreMode =
+        submitRequest.restoreMode != null && submitRequest.flinkVersion.checkVersion(
+          FlinkRestoreMode.SINCE_FLINK_VERSION)
+      if (enableRestoreMode) {
         flinkConfig.setString(FlinkRestoreMode.RESTORE_MODE, submitRequest.restoreMode.getName);
       }
     }
@@ -358,7 +357,7 @@ trait FlinkClientTrait extends Logger {
   }
 
   private[client] def getParallelism(submitRequest: SubmitRequest): Integer = {
-    if (submitRequest.properties.containsKey(KEY_FLINK_PARALLELISM())) {
+    if (submitRequest.hasProp(KEY_FLINK_PARALLELISM())) {
       Integer.valueOf(submitRequest.properties.get(KEY_FLINK_PARALLELISM()).toString)
     } else {
       getFlinkDefaultConfiguration(submitRequest.flinkVersion.flinkHome)

--- a/streampark-flink/streampark-flink-client/streampark-flink-client-core/src/main/scala/org/apache/streampark/flink/client/trait/FlinkClientTrait.scala
+++ b/streampark-flink/streampark-flink-client/streampark-flink-client-core/src/main/scala/org/apache/streampark/flink/client/trait/FlinkClientTrait.scala
@@ -20,9 +20,9 @@ package org.apache.streampark.flink.client.`trait`
 import org.apache.streampark.common.Constant
 import org.apache.streampark.common.conf.ConfigKeys._
 import org.apache.streampark.common.conf.Workspace
-import org.apache.streampark.common.enums.{ApplicationType, FlinkDevelopmentMode, FlinkExecutionMode, FlinkRestoreMode}
+import org.apache.streampark.common.enums._
 import org.apache.streampark.common.fs.FsOperator
-import org.apache.streampark.common.util.{AssertUtils, DeflaterUtils, ExceptionUtils, Logger, PropertiesUtils, SystemPropertyUtils, Utils}
+import org.apache.streampark.common.util._
 import org.apache.streampark.flink.client.bean._
 import org.apache.streampark.flink.core.FlinkClusterClient
 import org.apache.streampark.flink.core.conf.FlinkRunOption
@@ -80,10 +80,42 @@ trait FlinkClientTrait extends Logger {
          |-------------------------------------------------------------------------------------------
          |""".stripMargin)
 
+    // prepare flink config
+    val flinkConfig = prepareConfig(submitRequest)
+
+    // set JVMOptions..
+    setJvmOptions(submitRequest, flinkConfig)
+
+    setConfig(submitRequest, flinkConfig)
+
+    Try(doSubmit(submitRequest, flinkConfig)) match {
+      case Success(resp) => resp
+      case Failure(e) =>
+        logError(
+          s"flink job ${submitRequest.appName} start failed, " +
+            s"executionMode: ${submitRequest.executionMode.getName}, " +
+            s"detail: ${ExceptionUtils.stringifyException(e)}")
+        throw e
+    }
+  }
+
+  private[this] def prepareConfig(submitRequest: SubmitRequest): Configuration = {
+
     val (commandLine, flinkConfig) = getCommandLineAndFlinkConfig(submitRequest)
 
     submitRequest.developmentMode match {
       case FlinkDevelopmentMode.PYFLINK =>
+        val pythonVenv: String = Workspace.local.APP_PYTHON_VENV
+        AssertUtils.required(FsOperator.lfs.exists(pythonVenv), s"$pythonVenv File does not exist")
+
+        flinkConfig
+          // python.archives
+          .safeSet(PythonOptions.PYTHON_ARCHIVES, pythonVenv)
+          // python.client.executable
+          .safeSet(PythonOptions.PYTHON_CLIENT_EXECUTABLE, Constant.PYTHON_EXECUTABLE)
+          // python.executable
+          .safeSet(PythonOptions.PYTHON_EXECUTABLE, Constant.PYTHON_EXECUTABLE)
+
         val flinkOptPath: String = System.getenv(ConfigConstants.ENV_FLINK_OPT_DIR)
         if (StringUtils.isBlank(flinkOptPath)) {
           logWarn(s"Get environment variable ${ConfigConstants.ENV_FLINK_OPT_DIR} fail")
@@ -123,7 +155,7 @@ trait FlinkClientTrait extends Logger {
     }
 
     // set savepoint parameter
-    if (submitRequest.savePoint != null) {
+    if (StringUtils.isNotBlank(submitRequest.savePoint)) {
       flinkConfig.safeSet(SavepointConfigOptions.SAVEPOINT_PATH, submitRequest.savePoint)
       flinkConfig.setBoolean(
         SavepointConfigOptions.SAVEPOINT_IGNORE_UNCLAIMED_STATE,
@@ -135,20 +167,7 @@ trait FlinkClientTrait extends Logger {
       }
     }
 
-    // set JVMOptions..
-    setJvmOptions(submitRequest, flinkConfig)
-
-    setConfig(submitRequest, flinkConfig)
-
-    Try(doSubmit(submitRequest, flinkConfig)) match {
-      case Success(resp) => resp
-      case Failure(e) =>
-        logError(
-          s"flink job ${submitRequest.appName} start failed, " +
-            s"executionMode: ${submitRequest.executionMode.getName}, " +
-            s"detail: ${ExceptionUtils.stringifyException(e)}")
-        throw e
-    }
+    flinkConfig
   }
 
   private[this] def setJvmOptions(
@@ -276,27 +295,15 @@ trait FlinkClientTrait extends Logger {
 
       submitRequest.developmentMode match {
         case FlinkDevelopmentMode.PYFLINK =>
-          val pythonVenv: String = Workspace.local.APP_PYTHON_VENV
-
-          AssertUtils.required(
-            FsOperator.lfs.exists(pythonVenv),
-            s"$pythonVenv File does not exist")
-          flinkConfig
-            // python.archives
-            .safeSet(PythonOptions.PYTHON_ARCHIVES, pythonVenv)
-            // python.client.executable
-            .safeSet(PythonOptions.PYTHON_CLIENT_EXECUTABLE, Constant.PYTHON_EXECUTABLE)
-            // python.executable
-            .safeSet(PythonOptions.PYTHON_EXECUTABLE, Constant.PYTHON_EXECUTABLE)
           if (submitRequest.libs.nonEmpty) {
             // BUG: https://github.com/apache/incubator-streampark/issues/3761
             // builder.setUserClassPaths(Lists.newArrayList(submitRequest.libs: _*))
           }
         case _ =>
           builder
-            // BUG: https://github.com/apache/incubator-streampark/issues/3761
-            // .setUserClassPaths(Lists.newArrayList(submitRequest.classPaths: _*))
             .setJarFile(jarFile)
+        // BUG: https://github.com/apache/incubator-streampark/issues/3761
+        // .setUserClassPaths(Lists.newArrayList(submitRequest.classPaths: _*))
       }
       builder
     }
@@ -308,6 +315,7 @@ trait FlinkClientTrait extends Logger {
       getParallelism(submitRequest),
       null,
       false)
+
     packageProgram -> jobGraph
   }
 
@@ -344,9 +352,9 @@ trait FlinkClientTrait extends Logger {
   private[this] def getCustomCommandLines(flinkHome: String): JavaList[CustomCommandLine] = {
     val flinkDefaultConfiguration: Configuration = getFlinkDefaultConfiguration(flinkHome)
     // 1. find the configuration directory
-    val configurationDirectory = s"$flinkHome/conf"
+    val confDir = s"$flinkHome/conf"
     // 2. load the custom command lines
-    loadCustomCommandLines(flinkDefaultConfiguration, configurationDirectory)
+    loadCustomCommandLines(flinkDefaultConfiguration, confDir)
   }
 
   private[client] def getParallelism(submitRequest: SubmitRequest): Integer = {
@@ -367,23 +375,20 @@ trait FlinkClientTrait extends Logger {
     val cliArgs = {
       val optionMap = new mutable.HashMap[String, Any]()
       submitRequest.appOption
-        .filter {
+        .foreach {
           opt =>
             val verify = commandLineOptions.hasOption(opt._1)
             if (!verify) {
               logWarn(s"param:${opt._1} is error,skip it.")
-            }
-            verify
-        }
-        .foreach {
-          opt =>
-            val option = commandLineOptions.getOption(opt._1.trim).getOpt
-            Try(opt._2.toBoolean).getOrElse(opt._2) match {
-              case b if b.isInstanceOf[Boolean] =>
-                if (b.asInstanceOf[Boolean]) {
-                  optionMap += s"-$option" -> true
-                }
-              case v => optionMap += s"-$option" -> v
+            } else {
+              val option = commandLineOptions.getOption(opt._1.trim).getOpt
+              Try(opt._2.toBoolean).getOrElse(opt._2) match {
+                case b if b.isInstanceOf[Boolean] =>
+                  if (b.asInstanceOf[Boolean]) {
+                    optionMap += s"-$option" -> true
+                  }
+                case v => optionMap += s"-$option" -> v
+              }
             }
         }
 

